azavea/raster-vision
# Lazy Loading of TensorFlow and Keras

## Overview

These changes cause TensorFlow and Keras to be lazily-loaded.

Methodology:

~~The one top-level importation of tensorflow is now done lazily with `importlib` (see first commit and [this](https://stackoverflow.com/questions/42703908/how-do-i-use-importlib-lazyloader) stack overflow page).~~  The top-level tensorflow import has been turned into several local imports.  The top-level keras import has been turned into a local import because keras seems to not work with the `importlib` technique.  Finally, all `from ...` imports have been made local.

~~Getting rid of `importlib` entirely and turning the one top-level tensorflow import into several local imports is also possible direction.~~

To Test:

Run `python tests/core/test_box.py` and observe that output from TensorFlow is no longer produced.  The unit tests and integration tests should still succeed.

In a container, uninstall tensorflow and delete the `/opt/tf-models` directory, then run some tests that do no involve them:
```bash
rm -r /opt/tf-models
pip uninstall tensorflow
python tests/core/test_box.py
```

### Checklist


### Notes

Optional. Ancillary topics, caveats, alternative strategies that didn't work out, anything else.

## Testing Instructions

* How to test this PR
* Prefer bulleted description
* Start after checking out this branch
* Include any setup required, such as rebuilding the Docker image.
* Include test case, and expected output if not captured by automated tests.

Closes #388
